interpreter/llvm/src/tools/clang/include/clang/Serialization/ASTReader.h

```diff
@@ -1210,6 +1210,12 @@ class ASTReader
       : Mod(Mod), ImportedBy(ImportedBy), ImportLoc(ImportLoc) { }
   };
 
+  uint32_t getGenerationOrNull() const {
+    if (ContextObj)
```

Comment:
In which cases we have no `ContextObj`? I'd prefer to assert here if possible (and rename the routine).

Reply:
The ASTReader only has only an optional ASTContext, it also supposed to work without. You can also just read preprocessor information (e.g. ASTUnit is using this) and then you don't need to supply one. Other people might use it in other contexts where have no ASTContext.

The current clang code only allows increasing the generation counter with an ASTContext (as it only increases the generation counter of the top-most ASTContext source), so we anyway don't keep an generation counter without an ASTContext right now.

Reply:
We could rename it to getGeneration(), the `OrNull` is probably rather confusing.

Reply:
Can you try asserting and run the test suite. I am concerned about the case where we `return 0` but still add an entry to the table of `II`.

Reply:
I started a build with the added assert here: https://teemperor.de/ccir/git:RefactorGenerationCounterAssert

It will probably fail in the ASTUnit tests, but there we just have a 0 generation everywhere. That's what the code is doing.

Reply:
I see, it fails only for the cases where we are preprocessing modules. Maybe it'd be worth tweaking that assert. I'd be much more comfortable to control the number generation.